src/workerd/server/isolate-limit-enforcer.c++

```diff
@@ -0,0 +1,227 @@
+#include <kj/filesystem.h>
+#include <fcntl.h>
+
+namespace workerd {
+namespace {
+
+class NullIsolateLimitEnforcer final: public IsolateLimitEnforcer {
+public:
+  v8::Isolate::CreateParams getCreateParams() override { return {}; }
+  void customizeIsolate(v8::Isolate* isolate) override {}
+  ActorCacheSharedLruOptions getActorCacheLruOptions() override {
+    // TODO(someday): Make this configurable?
+    return {
+      .softLimit = 16 * (1ull << 20), // 16 MiB
+      .hardLimit = 128 * (1ull << 20), // 128 MiB
+      .staleTimeout = 30 * kj::SECONDS,
+      .dirtyListByteLimit = 8 * (1ull << 20), // 8 MiB
+      .maxKeysPerRpc = 128,
+
+      // For now, we use `neverFlush` to implement in-memory-only actors.
+      // See WorkerService::getActor().
+      .neverFlush = true
+    };
+  }
+  kj::Own<void> enterStartupJs(
+      jsg::Lock& lock, kj::Maybe<kj::Exception>& error) const override {
+    return {};
+  }
+  kj::Own<void> enterDynamicImportJs(
+      jsg::Lock& lock, kj::Maybe<kj::Exception>& error) const override {
+    return {};
+  }
+  kj::Own<void> enterLoggingJs(
+      jsg::Lock& lock, kj::Maybe<kj::Exception>& error) const override {
+    return {};
+  }
+  kj::Own<void> enterInspectorJs(
+      jsg::Lock& loc, kj::Maybe<kj::Exception>& error) const override {
+    return {};
+  }
+  void completedRequest(kj::StringPtr id) const override {}
+  bool exitJs(jsg::Lock& lock) const override { return false; }
+  void reportMetrics(IsolateObserver& isolateMetrics) const override {}
+  kj::Maybe<size_t> checkPbkdfIterations(jsg::Lock& lock, size_t iterations) const override {
+    // No limit on the number of iterations in workerd
+    return kj::none;
+  }
+};
+
+class LimitedArrayBufferAllocator final: public v8::ArrayBuffer::Allocator {
+public:
+  LimitedArrayBufferAllocator(size_t limit): limit(limit) {}
+  ~LimitedArrayBufferAllocator() {}
+
+  void* Allocate(size_t length) override {
+    if (length > limit) return nullptr;
+    return calloc(length, 1);
+  }
+  void* AllocateUninitialized(size_t length) override {
+    if (length > limit) return nullptr;
+    return malloc(length);
+  }
+  void Free(void* data, size_t length) override {
+    free(data);
+  }
+
+private:
+  size_t limit;
+};
+
+class ConfiguredIsolateLimitEnforcer final: public IsolateLimitEnforcer {
+public:
+  ConfiguredIsolateLimitEnforcer(server::config::Worker::Limits::Reader limits)
+      : softHeapLimitMb(limits.getHeapSoftLimitMb()),
+        heapSnapshotNearHeapLimit(limits.getHeapSnapshotNearHeapLimit()),
+        heapLimitMultiplier(limits.getHeapLimitMultiplier()),
+        heapLimitExceedsMax(limits.getHeapLimitExceedsMax()) {}
+
+  v8::Isolate::CreateParams getCreateParams() override {
+    v8::Isolate::CreateParams params;
+    uint64_t softLimit = softHeapLimitMb * 1024 * 1024;
+    if (softLimit > 0) {
+      params.constraints.set_max_young_generation_size_in_bytes(
+          kj::min(softLimit, 2 * 1024 * 1024));
```

Comment:
Should we make the `2` be configurable, too?

Reply:
I actually think we shouldn't. We shouldn't offer config options that no one actually wants or needs. Doing so only overwhelms the user and locks us into supporting those options forever.

Reply:
Well, I'm not sure we can yet say no one actually wants or needs it yet ;-) ... there's no rush to land this so we can look at various options. I'd like to get some feedback from the devprod folks before we lock down any of the configuration options here, as well as see if we can solicit some input from users.